modules/seqexec/server/src/main/scala/seqexec/server/tcs/ScienceFoldPositionCodex.scala

```diff
@@ -4,66 +4,52 @@
-import seqexec.server.tcs.TcsController.{LightSource, ScienceFoldPosition}
-import seqexec.server.tcs.TcsControllerEpics.InstrumentPorts
+import seqexec.server.tcs.TcsController.LightSource
+import seqexec.server.tcs.TcsControllerEpics.ScienceFold.{Parked, Position}
+import seqexec.server.tcs.TcsControllerEpics.ScienceFold
 
 // Decoding and encoding the science fold position require some common definitions, therefore I
 // put them inside an object
 private[server] object ScienceFoldPositionCodex {
 
   import LightSource._
-  import ScienceFoldPosition._
 
   private val AO_PREFIX = "ao2"
   private val GCAL_PREFIX = "gcal2"
   private val PARK_POS = "park-pos"
 
-  val BottomPort: Int = 1
-  val InvalidPort: Int = 0
-
-  def portFromSinkName(ports: InstrumentPorts)(n: LightSinkName): Option[Int] = {
-    val port = n match {
-      case LightSinkName.Gmos |
-           LightSinkName.Gmos_Ifu => ports.gmosPort
-      case LightSinkName.Niri_f6 |
-           LightSinkName.Niri_f14 |
-           LightSinkName.Niri_f32 => ports.niriPort
-      case LightSinkName.Nifs     => ports.nifsPort
-      case LightSinkName.Gnirs    => ports.gnirsPort
-      case LightSinkName.F2       => ports.flamingos2Port
-      case LightSinkName.Gpi      => ports.gpiPort
-      case LightSinkName.Ghost    => ports.ghostPort
-      case LightSinkName.Gsaoi    => ports.gsaoiPort
-      case LightSinkName.Ac |
-           LightSinkName.Hr       => BottomPort
-      case LightSinkName.Phoenix |
-           LightSinkName.Visitor  => InvalidPort
-    }
-    (port =!= InvalidPort).option(port)
-  }
-
   private def findSinkInSFName(str: String): Option[LightSinkName] =
     LightSinkName.all.find(i => str.startsWith(i.name))
 
-  implicit val decodeScienceFoldPosition: DecodeEpicsValue[String, Option[ScienceFoldPosition]] = DecodeEpicsValue(
+  private def findPortInSFName(str: String): Option[Int] = str.parseIntOption
+
+  implicit val decodeScienceFold: DecodeEpicsValue[String, Option[ScienceFold]] = DecodeEpicsValue(
     (t: String) => if (t.startsWith(PARK_POS)) Parked.some
-    else if (t.startsWith(AO_PREFIX))
-      findSinkInSFName(t.substring(AO_PREFIX.length)).map(Position(AO, _))
-    else if (t.startsWith(GCAL_PREFIX))
-      findSinkInSFName(t.substring(GCAL_PREFIX.length)).map(Position(GCAL, _))
-    else findSinkInSFName(t).map(Position(Sky, _))
+    else if (t.startsWith(AO_PREFIX)) for {
+      sink <- findSinkInSFName(t.substring(AO_PREFIX.length))
+      port <- findPortInSFName(t.substring(AO_PREFIX.length + sink.name.length))
+    } yield Position(AO, sink, port)
+    else if (t.startsWith(GCAL_PREFIX)) for {
+      sink <- findSinkInSFName(t.substring(GCAL_PREFIX.length))
+      port <- findPortInSFName(t.substring(GCAL_PREFIX.length + sink.name.length))
+    } yield Position(GCAL, sink, port)
+    else for {
+      sink <- findSinkInSFName(t)
+      port <- findPortInSFName(t.substring(sink.name.length))
+    } yield Position(Sky, sink, port)
```

Comment:
We have @tpolecat 's `Atto` for this kind of thing so we're not having to work with substrings and indices.  I'm not an expert and I haven't actually tried to compile, but it might go something like this:

```scala
import atto._, Atto._

val lightSink: Parser[LightSinkName] =
  many(letter).map { cs =>
    val sinkName = cs.mkString
    LightSinkName.all.find(_.name == sinkName)
  }.collect { case Some(lsn) => lsn }

def prefixed(p: String, s: LightSource): Parser[Position] =
  (string(p) ~> lightSink ~ int).map { case (ls, port) => Position(s, ls, port) }

val park: Parser[Position] =
  (string(PARK_POS) <~ many(anyChar)).as(Parked)

(park | prefixed(AO_PREFIX, AO) | prefixed(GCAL_PREFIX, GCAL) | prefixed("", SKY)).parseOnly(t).option
```

Reply:
Oh I didn't even look at what was happening in the `find...` methods. Yes this would be much better than mucking around with strings.